src/types.cpp

```diff
@@ -119,157 +119,140 @@ namespace Exiv2 {
+        if (pData_.size() < 2 || offset > (pData_.size() - 2)) {
             throw std::overflow_error("Overflow in Exiv2::DataBuf::write_uint16");
         }
         us2Data(&pData_[offset], x, byteOrder);
     }
 
     uint32_t Exiv2::DataBuf::read_uint32(size_t offset, ByteOrder byteOrder) const {
-        if (size_ < 4 || offset > static_cast<size_t>(size_ - 4)) {
+        if (pData_.size() < 4 || offset > (pData_.size() - 4)) {
             throw std::overflow_error("Overflow in Exiv2::DataBuf::read_uint32");
         }
         return getULong(&pData_[offset], byteOrder);
     }
 
     void Exiv2::DataBuf::write_uint32(size_t offset, uint32_t x, ByteOrder byteOrder) {
-        if (size_ < 4 || offset > static_cast<size_t>(size_ - 4)) {
+        if (pData_.size() < 4 || offset > (pData_.size() - 4)) {
             throw std::overflow_error("Overflow in Exiv2::DataBuf::write_uint32");
         }
         ul2Data(&pData_[offset], x, byteOrder);
     }
 
     uint64_t Exiv2::DataBuf::read_uint64(size_t offset, ByteOrder byteOrder) const {
-        if (size_ < 8 || offset > static_cast<size_t>(size_ - 8)) {
+        if (pData_.size() < 8 || offset > (pData_.size() - 8)) {
             throw std::overflow_error("Overflow in Exiv2::DataBuf::read_uint64");
         }
         return getULongLong(&pData_[offset], byteOrder);
     }
 
     void Exiv2::DataBuf::write_uint64(size_t offset, uint64_t x, ByteOrder byteOrder) {
-        if (size_ < 8 || offset > static_cast<size_t>(size_ - 8)) {
+        if (pData_.size() < 8 || offset > (pData_.size() - 8)) {
             throw std::overflow_error("Overflow in Exiv2::DataBuf::write_uint64");
         }
         ull2Data(&pData_[offset], x, byteOrder);
     }
 
     /// \todo do not use void*
     void Exiv2::DataBuf::copyBytes(size_t offset, const void* buf, size_t bufsize) {
-        if (static_cast<size_t>(size_) < bufsize || offset > size_ - bufsize) {
+        if (pData_.size() < bufsize || offset > pData_.size() - bufsize) {
             throw std::overflow_error("Overflow in Exiv2::DataBuf::copyBytes");
         } if (bufsize > 0) {
             memcpy(&pData_[offset], buf, bufsize);
         }
     }
 
     int Exiv2::DataBuf::cmpBytes(size_t offset, const void* buf, size_t bufsize) const {
-        if (static_cast<size_t>(size_) < bufsize || offset > size_ - bufsize) {
+        if (pData_.size() < bufsize || offset > pData_.size() - bufsize) {
             throw std::overflow_error("Overflow in Exiv2::DataBuf::cmpBytes");
         }
         return memcmp(&pData_[offset], buf, bufsize);
     }
 
     byte* Exiv2::DataBuf::data(size_t offset) {
-        if (static_cast<size_t>(size_) < offset) {
+        if (pData_.size() < offset) {
             throw std::overflow_error("Overflow in Exiv2::DataBuf::c_data");
-        } else if (size_ == 0 || static_cast<size_t>(size_) == offset) {
+        } else if (pData_.size() == 0 || pData_.size() == offset) {
```

Comment:
clang-tidy will throw a warning on this. It's better to use the empty() method instead of comparing to 0.